sklearn/preprocessing/_data.py

```diff
@@ -2595,15 +2595,13 @@ def _dense_fit(self, X, random_state):
         n_samples, n_features = X.shape
         references = self.references_ * 100
 
-        self.quantiles_ = []
-        for col in X.T:
-            if self.subsample < n_samples:
-                subsample_idx = random_state.choice(
-                    n_samples, size=self.subsample, replace=False
-                )
-                col = col.take(subsample_idx, mode="clip")
-            self.quantiles_.append(np.nanpercentile(col, references))
-        self.quantiles_ = np.transpose(self.quantiles_)
+        if self.subsample < n_samples:
+            subsample_idx = random_state.choice(
+                n_samples, size=self.subsample, replace=False
+            )
+            X = _safe_indexing(X, subsample_idx)
+
+        self.quantiles_ = np.nanpercentile(X, references, axis=0)
```

Comment:
Could you also factorize this part that is in common with `KBinsDiscretizer` by adding a new utility function called `subsample`.

We would need to add it in `classes.rst` and have a small test to check the behaviour as well.

Reply:
Ok, but I'm not sure which is better. Could you give me some suggestions?

Option 1

```python
# QuantileTransformer and KBinsDiscretizer
if self.subsample is not None and n_samples > self.subsample:
    X = subsample(X, n_samples=self.subsample, random_state=random_state)

# subsample method in utility
def subsample(*arrays, n_samples, random_state):
    return resample(*arrays, replace=False, n_samples=n_samples, random_state=random_state)
```
Option 2
```python
# QuantileTransformer and KBinsDiscretizer
X = subsample(X, n_samples=self.subsample, random_state=random_state)

# subsample method in utility
def subsample(*arrays, n_samples, random_state):
    subsample = n_samples
    first = arrays[0]
    n_samples = first.shape[0] if hasattr(first, "shape") else len(first)
    if subsample is not None and n_samples > subsample:
        return resample(*arrays, replace=False, n_samples=n_samples, random_state=random_state)
    else:
        return arrays
```

Reply:
Oh actually call directly the `resample` function. I did not know that it was exposing all the necessary parameters:

```python
if self.subsample is not None and n_samples > self.subsample:
    # Take a subsample of `X`
    X = resample(X, replace=False, n_samples=n_samples, random_state=random_state)
```

Reply:
done